### DISCUSSION TITLE
keycloak operator upgrade procedure

### DISCUSSION THREAD

**Discussion by rakeshreddyrg09** [keycloak/keycloak#22690] [Feedback]
Hello, fellow developers

I'm in the process of upgrading Keycloak from version 19.0.3 to 22.0.1 using the Keycloak Operator. I've explored a few options but have encountered some challenges. Could you kindly assist me in finding the best approach for a smooth upgrade?

Option 1 - Realm Export:
I've considered using the realm_export approach, but unfortunately, it doesn't provide an option to export users. I need a solution that ensures the smooth migration of both realms and users.

Option 2 - Database Mapping:
Another attempt involved creating a new PostgreSQL database using the existing database data. During the new Keycloak installation, I linked it to this PostgreSQL. However, I encountered an issue where Keycloak pod logs indicated that tables (client, realm, databasechangeloglock, databasechangelog) already exist. Deleting them resulted in further errors due to existing tables.

Option 3 - Extracting User Information:
I attempted to extract user information using the command:
kubectl exec -it -n test keycloak-0 -- /opt/jboss/keycloak/bin/kcadm.sh get users -r TEST_DEV --offset 0
This command provides user information but doesn't include assigned roles or attached groups. I need a way to capture complete user details including roles and groups.

My setup is based on EKS with the Keycloak Operator installed. I've applied CRDs ([keycloakrealmimport](http://kubectl%20apply%20-f%20https//raw.githubusercontent.com/keycloak/keycloak-k8s-resources/22.0.1/kubernetes/keycloakrealmimports.k8s.keycloak.org-v1.yml), [keycloak](http://kubectl%20apply%20-f%20https//raw.githubusercontent.com/keycloak/keycloak-k8s-resources/22.0.1/kubernetes/keycloaks.k8s.keycloak.org-v1.yml)) as specified in the provided link. [This](http://kubectl%20apply%20-f%20https//raw.githubusercontent.com/keycloak/keycloak-k8s-resources/22.0.1/kubernetes/kubernetes.yml) setup brought the operator and deployed Keycloak is [here](https://www.keycloak.org/operator/basic-deployment).

While I've explored the resources available, I've encountered challenges. I'm reaching out to the community for guidance and insights on the best approach to upgrade Keycloak successfully. Any help would be greatly appreciated. Thank you all in advance!

**Comment by rakeshreddyrg09** [comment]
kc bin in keycloak dir
![image](https://github.com/keycloak/keycloak/assets/48629246/3159cd06-7190-4ee6-8b56-a6ec3d6c3a65)
kc bin home dir
![image](https://github.com/keycloak/keycloak/assets/48629246/d6588f52-fc17-4d6e-aad4-0b192961eaac)
my keycloak version details
![image](https://github.com/keycloak/keycloak/assets/48629246/1c5fe64d-43e8-422e-89b4-409a8815958b)

**Comment by vmuzikar** [comment]
The upgrade should be seamless, DB migration will be performed automatically, no need to manually export/import anything. Just update the Operator, it will roll an update on your deployment.
